packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewer/PdfViewerTextLayer.tsx

```diff
@@ -0,0 +1,179 @@
+import React, { FC, useEffect, useRef, useCallback } from 'react';
+import cx from 'classnames';
+import { PDFPageProxy, PDFPageViewport, TextContent, TextContentItem } from 'pdfjs-dist';
+import { EventBus } from 'pdfjs-dist/lib/web/ui_utils';
+import { TextLayerBuilder } from 'pdfjs-dist/lib/web/text_layer_builder';
+import useAsyncFunctionCall from './useAsyncFunctionCall';
+
+interface Props {
+  className?: string;
+
+  /**
+   * PDF page from pdfjs
+   */
+  loadedPage: PDFPageProxy | null | undefined;
+
+  /**
+   * Zoom factor, where `1` is equal to 100%
+   */
+  scale: number;
+
+  /**
+   * Callback for text layer info
+   */
+  setRenderedText?: (info: PdfRenderedText | null) => any;
+}
+
+export type PdfRenderedText = {
+  /**
+   * PDF text content
+   */
+  textContent: TextContent & {
+    styles: { [styleName: string]: CSSStyleDeclaration };
+  };
+
+  /**
+   * Text span DOM elements rendered on the text layer
+   */
+  textDivs: HTMLElement[];
+
+  /**
+   * Pdf page viewport used to render text items
+   */
+  viewport: PDFPageViewport;
+
+  /**
+   * Page number, starting at 1
+   */
+  page: number;
+};
+
+const PdfViewerTextLayer: FC<Props> = ({
+  className,
+  loadedPage,
+  scale = 1,
+  setRenderedText = () => {}
+}) => {
+  const textLayerRef = useRef<HTMLDivElement>(null);
+  const textLayerDiv = textLayerRef.current;
+
+  // load text content from the page
+  const loadedText = useAsyncFunctionCall(
+    useCallback(async () => {
+      if (loadedPage) {
+        const viewport = loadedPage.getViewport({ scale });
+        const textContent = await loadedPage.getTextContent();
+        return { textContent, viewport, page: loadedPage.pageNumber, scale };
+      }
+      return null;
+    }, [loadedPage, scale])
+  );
+
+  // render text content
+  const renderedText = useAsyncFunctionCall(
+    useCallback(
+      async (signal: AbortSignal) => {
+        if (textLayerDiv && loadedText) {
+          const { textContent, viewport, scale, page } = loadedText;
+
+          const builder = new TextLayerBuilder({
+            textLayerDiv,
+            viewport,
+            eventBus: new EventBus(),
+            pageIndex: page - 1
+          });
+          signal.addEventListener('abort', () => builder.cancel());
```

Comment:
I'm really liking this AbortController usage pattern. We should consider bringing this in to main Tooling (as an import from this library).